dotnet/src/Connectors/Connectors.OpenAI/Core/OpenAIFunction.cs

```diff
@@ -127,52 +139,143 @@ internal OpenAIFunction(
-    private static KernelJsonSchema GetDefaultSchemaForTypelessParameter(string? description)
+    private static KernelJsonSchema GetDefaultSchemaForTypelessParameter(string? description, bool allowStrictSchemaAdherence)
     {
         // If there's a description, incorporate it.
         if (!string.IsNullOrWhiteSpace(description))
         {
-            return KernelJsonSchemaBuilder.Build(typeof(string), description);
+            return allowStrictSchemaAdherence ?
+                GetOptionalStringSchemaWithDescription(description!) :
+                KernelJsonSchemaBuilder.Build(typeof(string), description, AIJsonSchemaCreateOptions.Default);
         }
 
         // Otherwise, we can use a cached schema for a string with no description.
-        return s_stringNoDescriptionSchema;
+        return allowStrictSchemaAdherence ? s_stringNoDescriptionSchemaAndNull : s_stringNoDescriptionSchema;
+    }
+
+    private static KernelJsonSchema GetOptionalStringSchemaWithDescription(string description)
+    {
+        var jObject = new JsonObject
+        {
+            { "description", description },
+            { "type", new JsonArray { "string", "null" } },
+        };
+        return KernelJsonSchema.Parse(jObject.ToString());
+    }
+    private static KernelJsonSchema? GetSanitizedSchemaForStrictMode(KernelJsonSchema? schema, bool insertNullType)
+    {
+        if (schema is null)
+        {
+            return null;
+        }
+        var forbiddenPropertyNames = s_forbiddenKeywords.Where(k => schema.RootElement.TryGetProperty(k, out _)).ToArray();
+
+        if (forbiddenPropertyNames.Length > 0 || insertNullType && schema.RootElement.TryGetProperty(TypeKey, out var typeElement) &&
+            (typeElement.ValueKind == JsonValueKind.Array && !typeElement.EnumerateArray().Any(static t => NullType.Equals(t.GetString(), StringComparison.OrdinalIgnoreCase)) ||
+            typeElement.ValueKind == JsonValueKind.String && typeElement.GetString() != NullType))
+        {
+            var originalSchema = JsonSerializer.Serialize(schema.RootElement);
+            var parsedJson = JsonNode.Parse(originalSchema);
+
+            if (parsedJson is null)
+            {
+                return schema;
+            }
+
+            var jsonObject = parsedJson.AsObject();
+            foreach (var forbiddenPropertyName in forbiddenPropertyNames)
+            {
+                jsonObject.Remove(forbiddenPropertyName);
+            }
+
+            InsertNullTypeIfRequired(insertNullType, jsonObject);
+
+            return KernelJsonSchema.Parse(jsonObject.ToString());
+        }
+        return schema;
+    }
+    // https://platform.openai.com/docs/guides/structured-outputs#some-type-specific-keywords-are-not-yet-supported
+    private static void InsertNullTypeIfRequired(bool insertNullType, JsonObject jsonObject)
+    {
+        if (insertNullType && jsonObject.TryGetPropertyValue(TypeKey, out var typeValue))
```

Comment:
Azure OpenAI GPT-4 still sets parameters to default values even though null is an option.
OpenAI GPT-4 complains about the type being missing.

Reply:
The default behaviour was because my instance of Azure OpenAI was behind! It now works great!
I'm not sure why OpenAI GPT-4 is complaining. 4o isn't.